src/USVPlugin.cpp

```diff
@@ -19,27 +20,26 @@ void USVPlugin::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
     sdf::ElementPtr thrustersPluginElement = utilities::getPluginElementByName(
         modelSDF, "thrusters"
     );
-    // Since the wind plugin is optional and uses the same filename as the thrusters plugin, it should be found by name and be ignored if not found 
+
+    // Since the following plugins are optional and uses the same filename as the thrusters plugin,
+    // it should be found by name and be ignored if not found
     sdf::ElementPtr windPluginElement = utilities::findPluginElementByName(modelSDF, "wind_dynamics");
+    sdf::ElementPtr directForcePluginElement = utilities::findPluginElementByName(modelSDF, "direct_force");
 
     // Initialize communication node and subscribe to gazebo topic
     mNode = transport::NodePtr(new transport::Node());
     mNode->Init();
 
     mActuators = new Actuators(mModel, mNode);
-    mThrusters = new Thrusters;
 
     mWorldUpdateEvent = event::Events::ConnectWorldUpdateBegin(
         boost::bind(&USVPlugin::updateBegin, this, _1)
     );
 
-    loadThrusters(thrustersPluginElement);
-    loadRudders(thrustersPluginElement);
-
-    if (windPluginElement) {
-        mWind = new Wind;
-        loadWindParameters(windPluginElement);
-    }
+    loadThrusters(thrustersPluginElement, mThrusters);
+    loadRudders(thrustersPluginElement, mRudders);
+    loadWindParameters(windPluginElement, mWind);
+    loadDirectForceApplicationParameters(directForcePluginElement, mDirectForce);
```

Comment:
it fits (somewhat) into:

> The only valid case where passing an argument as non-const reference is when the caller wants the method to modify a value in-place, and copying the data from the method’s return value to the data structure to be modified would be too costly.

Also I would prefer to have a single syntax for loading plugins, and since `mRudders` is a vector, the best way of modifying it would be using an output argument

Reply:
I actually don't agree with you @caioaamaral 

For the vectors: copying a, what, max. 10 elements vector (rudders and thrusters on a system), once per initialization, is NOT a valid case of "too costly". I like the rationale for the change, but unless it's not possible (because e.g. Rudder would be non-copyable, but I think it is), the following would fit the standard better:

~~~
mThrusters = loadThrusters(thrustersPluginElement);
~~~

For the rest: assuming we can make the objects null objects by default (i.e. objects that will essentially do nothing when called), and copyable, we can avoid the pointers and just assign them (a Very Good Thing !).

In any case, what you do is plain buggy. You pass null pointers as what you believe are out arguments (but they are not since you don't pass them as reference !). And since you essentially set a null pointer to a non-null one, just returning the allocated value would be a lot clearer.

More generally speaking, a corollary of the "avoid out-arguments as much as possible", is not a pass to "let's change 5 methods to  out arguments because one would need it, for consistency reasons". It's rather "I'm sorry this method is different from the others, but that was needed because of the out argument".

We can have a more in-depth discussion about the whole of this change you made, but given what it is right now, I would very much prefer you limit yourself to the task at hand, keep that cleanup for later. What you propose here is definitely not better than what we had IMO. You had a good urge (the current code is far from great), but not this way.

Reply:
> For the vectors: copying a, what, max. 10 elements vector (rudders and thrusters on a system), once per initialization, is NOT a valid case of "too costly".

I was already considering any container copy "too costly", but I can see your point